src/cls/IPM/General/History.cls

```diff
@@ -0,0 +1,524 @@
+
+/// Moniker of the repository where the package is being installed. For load and uninstall, this will be empty.
+Property SourceMoniker As %IPM.DataType.RepoMoniker(MAXLEN = 100);
+
+/// Details of the repository where the package is being installed. For load and uninstall, this will be empty.
+Property SourceDetails As %String(MAXLEN = "");
+
+/// Time when the action started
+Property TimeStart As %TimeStamp [ Required ];
+
+/// Time when the action ended. This will be empty if the action is still in progress or ended abnormally.
+Property TimeEnd As %TimeStamp;
+
+/// Namespace where the action is being performed
+Property NameSpace As %String [ InitialExpression = {$NAMESPACE}, Required ];
+
+/// User who initiated the action
+Property UserName As %String [ InitialExpression = {$USERNAME}, Required ];
+
+/// Status of the action. If the action is still in progress, this will be 0. Otherwise, it will be the status code of the action.
+Property Success As %Status [ InitialExpression = 0, Required ];
+
+/// Whether the action is committed. In non developer mode, this indicates whether the action was successful. In developer mode, this is always true.
+Property Committed As %Boolean [ InitialExpression = 0, Required ];
+
+/// The command string that triggered the action
+Property CommandString As %String(MAXLEN = 8192) [ Required ];
+
+ClassMethod Init(Action As %String, Package As %IPM.DataType.ModuleName) As %IPM.General.History [ Private ]
+{
+    Set log = ..%New()
+    Set log.TimeStart = $ZDateTime($Now(), 3)
+    Set log.Action = Action
+    Set log.Package = Package
+    Set log.CommandString = $Get($$$ZPMCommandToLog, "<unset>")
+    Set log.NameSpace = $Namespace
+    Set log.UserName = $Username
+    $$$ThrowOnError(log.SetVersion()) // Set the version to placeholder 0.0.0-0, Will be updated before finalizing
+    $$$ThrowOnError(log.%Save()) // Save it now in case Finalize isn't called
+    Quit log
+}
+
+ClassMethod InstallInit(Package As %IPM.DataType.ModuleName) As %IPM.General.History
+{
+    Quit ..Init("install", Package)
+}
+
+ClassMethod LoadInit() As %IPM.General.History
+{
+    // Package name is not known at this point, so use a placeholder
+    Quit ..Init("load", "")
+}
+
+ClassMethod UninstallInit(Package As %IPM.DataType.ModuleName) As %IPM.General.History
+{
+    Quit ..Init("uninstall", Package)
+}
+
+Method SetSource(ServerName As %String) As %Status
+{
+    $$$ThrowOnError(..%Reload())
+    Set server = ##class(%IPM.Repo.Definition).ServerDefinitionKeyOpen(ServerName)
+    If '$IsObject(server) {
+        Quit $$$ERROR($$$GeneralError,$$$FormatText("Repository '%1' is not defined.", ServerName))
+    }
+    Set ..SourceName = ServerName
+    Set ..SourceMoniker = server.#MONIKER
+    Set ..SourceDetails = server.Details
+    Quit ..%Save()
+}
+
+Method SetVersion(pVersion As %IPM.General.SemanticVersion) As %Status
+{
+    If ..%Id() '= "" {
+        $$$ThrowOnError(..%Reload())
```

Comment:
For my understanding, why is an ID check needed here but not in SetName? If no particular reason, can call Reload() without the check in both. If there is a reason, please add a comment